### ISSUE TITLE
garbage occurs too often.

### ISSUE THREAD

**Issue by JeHwanYoo** [aivis/user-stream#8]
{"created_at":"Sat Dec 17 08:37:10 +0000 2016","id":810041133732597760,"id_str":"810041133732597760","text":"garbage","source":"\u003ca href=\"http:\/\/twitter.com\" rel=\"nofollow\"\u003eTwitter Web Client\u003c\/a\u003e","truncated":false,"in_reply_to_status_id":null,"in_reply_to_status_id_str":null,"in_reply_to_user_id":null,"in_reply_to_user_id_str":null,"in_reply_to_screen_name":null,"user":{"id":794036182753513472,"id_str":"794036182753513472","name":"Tweetaster","screen_name":"Tweetaster","location":"\ub300\ud55c\ubbfc\uad6d","url":null,"description":null,"protected":false,"verified":false,"followers_count":6,"friends_count":9,"listed_count":0,"favourites_count":9,"statuses_count":70,"created_at":"Thu Nov 03 04:39:13 +0000 2016","utc_offset":32400,"time_zone":"Seoul","geo_enabled":true,"lang":"ko","contributors_enabled":false,"is_translator":false,"profile_backgroun (The data is truncated here... so, it goes to garbage.)

Is it twitter's problem? or is it resolvable?

**Comment by JeHwanYoo** [comment]
I found a reason. When a data is coming from Twitter, it is sometimes  divided into two parts.
![image](https://cloud.githubusercontent.com/assets/13535954/21285650/df362d7a-c481-11e6-923d-0d0e5fdf27ed.png)

Is it only my problem? Does anyone have the same symptoms as me?

**Comment by JeHwanYoo** [comment]
The second coming data will not emit. This is because 'next_data_length' is NaN.

**Comment by JeHwanYoo** [comment]
Oh, now that I've checked, the source code is different than your git. I got it at npm. What the ...?

**Comment by JeHwanYoo** [comment]
This is from npm...
Sorry... But please check your npm.

```
var oauth   = require('oauth'),
    events  = require('events'),
    util    = require("util");

var user_stream_url     = 'https://userstream.twitter.com/1.1/user.json',
    request_token_url   = 'https://api.twitter.com/oauth/request_token',
    access_token_url    = 'https://api.twitter.com/oauth/access_token';

module.exports = Stream;

function Stream(params) {

    console.log('Success to Stream');

    if (!(this instanceof Stream)) {
        return new Stream(params);
    }

    events.EventEmitter.call(this);

    this.params = params;

    this.oauth = new oauth.OAuth(
        request_token_url,
        access_token_url,
        this.params.consumer_key,
        this.params.consumer_secret,
        '1.0',
        null,
        'HMAC-SHA1',
        null,
        {
          'Accept': '*/*',
          'Connection'
          : 'close',
          'User-Agent': 'user-stream.js'
        }
    );

}

//inherit
util.inherits(Stream, events.EventEmitter);

/**
 * Create twitter use stream
 *
 * Events:
 * - data
 * - garbage
 * - close
 * - error
 * - connected
 * - heartbeat
 *
 */
Stream.prototype.stream = function(params) {

    var stream = this;

    if (typeof params != 'object') {
        params = {};
    }

    //required params for lib
    params.delimited = 'length';
    params.stall_warnings = 'true';

    var request = this.oauth.post(
        user_stream_url,
        this.params.access_token_key,
        this.params.access_token_secret,
        params,
        null
    );

    /**
     * Destroy socket
     */
    this.destroy = function() {

        request.abort();

    }

    request.on('response', function(response) {

        // Any response code greater then 200 from steam API is an error
        if(response.statusCode > 200) {

            stream.emit('error', {type: 'response', data: {code:response.statusCode}});

        } else {
            var buffer = '',
                next_data_length = 0,
                end = '\r\n';

            //emit connected event
            stream.emit('connected');

            //set chunk encoding
            response.setEncoding('utf8');

            response.on('data', function(chunk) {

                //is heartbeat?
                if (chunk == end) {
                    stream.emit('heartbeat');
                    return;
                }

                //check whether new incomming data set
                if (!buffer.length) {
                    //get length of incomming data
                    var line_end_pos = chunk.indexOf(end);
                    next_data_length = parseInt(chunk.slice(0, line_end_pos));
                    //slice data length string from chunk
                    chunk = chunk.slice(line_end_pos+end.length);
                }

                // console.log(line_end_pos);
                // console.log(next_data_length);

                if (buffer.length != next_data_length) {
                    //data set recieved
                    //first remove end and append to buffer
                    buffer+= chunk.slice(0, chunk.indexOf(end));
                    //parse json
                    var parsed = false;
                    try {
                        //try parse & emit
                        buffer = JSON.parse(buffer);
                        parsed = true;
                    } catch(e) {
                        console.log('buffer is', buffer);
                        stream.emit('garbage', buffer);
                    }

                    //don't emit into "try" and emit only if data formatted
                    if (parsed) {
                        // console.log('get from twitter: ', buffer);
                        stream.emit('data', buffer);
                    }

                    //clean buffer
                    buffer = '';

                } else {
                    //append to buffer
                    buffer+=chunk;
                }

            });

            response.on('error', function(error) {

                stream.emit('close', error);

            });

            response.on('end', function() {

                stream.emit('close', 'socket end');

            });


            response.on('close', function() {

                request.abort();

            });
        }

    });

    request.on('error', function(error) {

        stream.emit('error', {type: 'request', data: error});

    });

    request.end();

}
```

**Comment by aivis** [comment]
Hi @JeHwanYoo 

I've published a new release `0.0.8` and it should include latest chunk aggregation changes: https://github.com/Roach/user-stream/blob/65069d5a7947223992bd1d964575f22825d3bdff/lib/user-stream.js
